File: utbot-framework/src/main/java/org/utbot/quickcheck/internal/generator/MarkerInterfaceGenerator.java


package org.utbot.quickcheck.internal.generator;

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.plugin.api.UtModel;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.Generator;
import org.utbot.quickcheck.internal.DefaultMethodHandleMaker;
import org.utbot.quickcheck.random.SourceOfRandomness;

import java.lang.invoke.MethodHandle;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

import static java.lang.System.identityHashCode;
import static java.lang.reflect.Proxy.newProxyInstance;
import static org.utbot.external.api.UtModelFactoryKt.classIdForType;

public class MarkerInterfaceGenerator<T> extends Generator<T> {
    private final Class<T> markerType;

    MarkerInterfaceGenerator(Class<T> markerType) {
        super(markerType);

        this.markerType = markerType;
    }

    @Override public UtModel generate(
        SourceOfRandomness random,
        GenerationStatus status) {
        return UtModelGenerator.getUtModelConstructor().construct(markerType.cast(
                newProxyInstance(
                        markerType.getClassLoader(),
                        new Class<?>[] { markerType },
                        new MarkerInvocationHandler<>(markerType))), classIdForType(markerType));
//        return markerType.cast(
//            newProxyInstance(
//                markerType.getClassLoader(),
//                new Class<?>[] { markerType },
//                new MarkerInvocationHandler<>(markerType)));
    }

    private static class MarkerInvocationHandler<T>
        implements InvocationHandler {

        private final Class<T> markerType;
        private final DefaultMethodHandleMaker methodHandleMaker =
            new DefaultMethodHandleMaker();

        MarkerInvocationHandler(Class<T> markerType) {
            this.markerType = markerType;
        }

        @Override
        public Object invoke(Object proxy, Method method, Object[] args)
            throws Throwable {

            if (Object.class.equals(method.getDeclaringClass()))
                return handleObjectMethod(proxy, method, args);
            if (method.isDefault())
                return handleDefaultMethod(proxy, method, args);

            return null;
        }

        private Object handleObjectMethod(
            Object proxy,
            Method method,
            Object[] args) {

            if ("equals".equals(method.getName()))
                return proxy == args[0];
            if ("hashCode".equals(method.getName()))
                return identityHashCode(proxy);

            return handleToString();
        }

        private Object handleDefaultMethod(
            Object proxy,
            Method method,
            Object[] args)
            throws Throwable {

            MethodHandle handle =
                methodHandleMaker.handleForSpecialMethod(method);
            return handle.bindTo(proxy).invokeWithArguments(args);
        }

        private String handleToString() {
            return "a synthetic instance of " + markerType;
        }
    }
}

File: utbot-greyboxfuzzer/src/main/kotlin/org/utbot/greyboxfuzzer/quickcheck/internal/generator/MarkerInterfaceGenerator.kt
package org.utbot.greyboxfuzzer.quickcheck.internal.generator

import org.utbot.greyboxfuzzer.quickcheck.generator.GeneratorContext
import org.utbot.greyboxfuzzer.util.classIdForType
import org.utbot.framework.plugin.api.UtModel
import org.utbot.greyboxfuzzer.quickcheck.generator.GenerationStatus
import org.utbot.greyboxfuzzer.quickcheck.generator.Generator
import org.utbot.greyboxfuzzer.quickcheck.internal.DefaultMethodHandleMaker
import org.utbot.greyboxfuzzer.quickcheck.random.SourceOfRandomness
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy

class MarkerInterfaceGenerator(private val markerType: Class<*>) : Generator(markerType) {
    override fun generate(
        random: SourceOfRandomness,
        status: GenerationStatus
    ): UtModel {
        return generatorContext.utModelConstructor.construct(
            markerType.cast(
                Proxy.newProxyInstance(
                    markerType.classLoader, arrayOf(markerType),
                    MarkerInvocationHandler(markerType)
                )
            ),
            classIdForType(markerType)
        )
    }

    private class MarkerInvocationHandler(private val markerType: Class<*>) : InvocationHandler {
        private val methodHandleMaker = DefaultMethodHandleMaker()

        override fun invoke(proxy: Any, method: Method, args: Array<Any>): Any? {
            if (Any::class.java == method.declaringClass) return handleObjectMethod(proxy, method, args)
            return if (method.isDefault) handleDefaultMethod(proxy, method, args) else null
        }

        private fun handleObjectMethod(
            proxy: Any,
            method: Method,
            args: Array<Any>
        ): Any {
            if ("equals" == method.name) return proxy === args[0]
            return if ("hashCode" == method.name) System.identityHashCode(proxy) else handleToString()
        }

        private fun handleDefaultMethod(
            proxy: Any,
            method: Method,
            args: Array<Any>
        ): Any {
            val handle = methodHandleMaker.handleForSpecialMethod(method)
            return handle.bindTo(proxy).invokeWithArguments(*args)
        }

        private fun handleToString(): String {
            return "a synthetic instance of $markerType"
        }
    }
}